gossip/api/types.go

```diff
@@ -0,0 +1,229 @@
+	"fmt"
+
+	"github.com/google/certificate-transparency-go/tls"
+)
+
+// PathPrefix is the common prefix for all HTTP entrypoints.
+const PathPrefix = "/gossip/v0/"
+
+// Add Log Head to Hub
+// POST https://<hub server>/gossip/v0/add-log-head
+// Inputs:
+//    source_url: The URL from which the log head was obtained
+//    head_data: The binary-encoded head
+//    log_signature: The source log's signature over head_data.
+// Outputs:
+//    timestamped_entry: TLS-encoded TimestampedEntry.
+//    hub_signature: hub's signature gossip_timestamped_entry
+
+// AddLogHeadPath is the final path component for this entrypoint.
+const AddLogHeadPath = "add-log-head"
+
+// AddLogHeadRequest represents the JSON request body sent to the add-log-head POST method.
+type AddLogHeadRequest struct {
+	SourceURL    string `json:"source_url"`
+	HeadData     []byte `json:"head_data"`
+	LogSignature []byte `json:"log_signature"`
+}
+
+// AddLogHeadResponse represents the JSON response body for the add-log-head POST method.
+type AddLogHeadResponse struct {
+	// TimestampedEntryData holds the TLS-encoding of a TimestampedEntry.
+	TimestampedEntryData []byte `json:"timestamped_entry"`
+	// HubSignature is a signature over GossipTimestamp.
+	HubSignature []byte `json:"hub_signature"`
+}
+
+// TimestampedEntry holds a timestamped entry in the hub, in a form suitable for TLS-encoding.
+// RFC 5246 notation:
+//   struct {
+//      opaque source_url<1,65535>;
+//      opaque head_data<1,65535>;
+//      opaque log_signature<1,65535>;
+//      uint64 hub_timestamp;
+//   } TimestampedEntry;
+type TimestampedEntry struct {
+	SourceURL    []byte `tls:"minlen:1,maxlen:65535"`
+	HeadData     []byte `tls:"minlen:1,maxlen:65535"`
+	LogSignature []byte `tls:"minlen:1,maxlen:65535"`
+	HubTimestamp uint64 // in nanoseconds since UNIX epoch
+}
+
+// SignedGossipTimestamp holds a timestamped entry in the gossip hub together with a
+// signature from the hub.  RFC 5246 notation:
+//   struct {
+//      TimestampedEntry timestamped_entry;
+//      opaque hub_signature<1,65535>;
+//   } SignedGossipTimestamp;
+type SignedGossipTimestamp struct {
+	TimestampedEntry TimestampedEntry
+	// HubSignature is over the TLS-encoding of GossipTimestamp.
+	HubSignature []byte `tls:"minlen:1,maxlen:65535"`
+}
+
+// Retrieve Latest Signed Tree Head for Hub
+// GET https://<hub server>/gossip/v0/get-sth
+// Inputs: None
+// Outputs:
+//    head_data: TLS-encoded HubTreeHead.
+//    hub_signature: signature over head_data.
+
+// GetSTHPath is the final path component for this entrypoint.
+const GetSTHPath = "get-sth"
+
+// GetSTHResponse represents the JSON response to the get-sth GET method.
+type GetSTHResponse struct {
+	// HeadData holds the TLS-encoding of a HubTreeHead.
+	TreeHeadData []byte `json:"head_data"`
+	// HubSignature is a signature over HeadData from the hub.
+	HubSignature []byte `json:"hub_signature"`
+}
+
+// HubTreeHead describes a head of the Hub's Merkle tree. RFC 5246 notation:
```

Comment:
Shouldn't this be the same format as the TrillianLogRoot?

Reply:
This works as long as the hub doesn't need to parse `TreeHeadData`, but in the long term it might need to, in which case we'll need to add this format to one of the deserialization options (aka. `types.LogRootV2`)

Reply:
I think the Hub would only need to parse `api.TreeHeadData` if we get all recursive (hubs logging STHs from hubs not CT Logs)&hellip;